### ISSUE TITLE
[unified-signatures][Feature-Request] Allow non-unified signatures if JSDoc is present

### ISSUE THREAD

**Issue by manuth** [typescript-eslint/typescript-eslint#1901]
**Repro**
1. Create a class with a method which has a small number of overloads
2. Create JSDocs for the overloads
3. Notice that a `unified-signature` failure is reported

```JSON
{
  "rules": {
    "@typescript-eslint/unified-signatures": ["warn"]
  }
}
```

```TS
class ResourceManager {
    /**
     * Gets a resource by its name.
     *  @param name The name of the resource to get.
     */
    public Get(name: string);
    /**
     * Gets a resource by its id.
     *  @param id The id of the resource to get.
     */
    public Get(id: number);
    /**
     * Gets a resource by its property-path.
     * @param path The property-path of the resource to get.
     */
    public Get(path: PropertyPath);
    public Get(entity: string | number | PropertyPath) { }
}
```

**Expected Result**
No `unified-signatures` failure should be reported as the signatures are distinguishable by their jsdoc.

**Actual Result**
An `unified-signatures`-failure is raised.

**Additional Info**
I think a method is easier to use when there is a good jsdoc on how to use it and what parameters to specify.
The above example is ways easier to use and to understand than, let's say, this here:
```ts
class ResourceManager {
    /**
     * Gets a resource.
     * @param entity The name, id or propertypath of the resource to get.
     */
    public Get(entity: string | number | PropertyPath) { }
}
```
Especially if your method-overloads have more than just one parameter.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.27.0` |
| `@typescript-eslint/parser`        | `2.27.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.4.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
Thanks for the request!

We don't do, and won't do JSDoc comment parsing - it's a big dependency to bring in and maintain, which makes this request a non-starter.
It's not possible to do this without JSDoc parsing as well, because otherwise slight changes between two comments (like an extra space or newline) would cause the two to no longer match, and cause a false negative.

But putting that aside - what would be the determining factor to differentiate the methods to make them "non-unifiable?" using JSDoc?

If you compare the `@param` tags, the differences will be the param name and description.
The param name will differ in the code anyways, so then it's just the param description that's different.
Using the param description to differentiate the two isn't a great thing, because it would mean that any typo would cause two otherwise identical signatures to not be reported by this rule:

```ts
/**
 * @param id some description that i misssssspelled oops
 */
public Get(id: string) {}
/**
 * @param id some description that i misspelled oops
 */
public Get(id: number) {}
```

This would be frustrating for an end user, esp when it can be v easy to introduce typos.

The same argument stands for using the parameter names themselves within the method signature - any slight misspelling would cause a false-negative.

```ts
public Get(id: string) {}
public Get(idd: number) {} // no longer unifiable
```

The point I'm trying to get at is that variable names and descriptions are human consumed _natural language_, so doing a naive `===` check isn't a good way for an automated tool to compare them.

Types and method signatures, on the other hand, are intended to be machine consumable - so there's no chance of a minor typo causing a false-negative (it would have to be major typo otherwise known as a bug :smile:)

----

If you want to keep your signatures non-unified, then you can use a disable comment to turn off the rule in that case.

**Comment by manuth** [comment]
My plan would be that just the presence of a JSDoc comment would make an overload non-unifiable.
Though as I can see JSDoc-comments aren't present in the AST tree, right?

**Comment by bradzacher** [comment]
They're present and queryable, but the presence of a JSDoc comment shouldn't disable this rule.

Typescript does not share JSDoc between overloads, so it's pretty common in many codebases to just copy+paste the JSDoc along with the signature, so that you always have a description of the function no matter the signature you're using, eg:

```ts
/**
 * This has a jsdoc
 */
function foo(arg: number): number;
// no jsdoc
function foo(arg: string): string;
/**
 * This also has a jsdoc, but typescript ignores it
 */
function foo(arg: any): any { }
```

![image](https://user-images.githubusercontent.com/7462525/79277779-59189e80-7e5f-11ea-8fb1-8e3d666c8510.png)
![image](https://user-images.githubusercontent.com/7462525/79277789-5f0e7f80-7e5f-11ea-903b-5b180d7c69b3.png)


So purely using the existence of JSDoc is not a good qualifier for most codebases, as there's a good chance are they'll be exactly the same. But as mentioned - doing an `===` check on the comment text isn't good, because even one extra erroneous space will cause them to not match.

**Comment by manuth** [comment]
Well you see - my thought is that the absence of a jsdoc would make un-unified overloads quite useless, wouldn't it...?
But of course I can just get along with disabling said rule, that's fine too, ofc.
